pydis_site/apps/api/viewsets/bot/nomination.py

```diff
@@ -1,32 +1,224 @@
+        - A user already has an active nomination;
+        - The `user` or `actor` are unknown to the site;
+        - The request contained a field that cannot be set at creation.
+
+    ### PATCH /bot/nominations/<id:int>
+    Update the nomination with the given `id` and return the updated nomination.
+    For active nominations, only the `reason` may be updated; for inactive
+    nominations, both the `reason` and `unnominate_reason` may be updated.
+
+    #### Request body
+    >>> {
+    ...     'reason': 'He would make a great helper',
+    ...     'unnominate_reason': 'He needs some time to mature his Python knowledge'
+    ... }
+
+    ### Response format
+    See `GET /bot/nominations/<id:int>`
+
+    ## Status codes
+    - 200: returned on success
+    - 400: if a field in the request body is invalid or disallowed
+    - 404: if an infraction with the given `id` could not be found
+
+    ### PATCH /bot/nominations/<id:int>/end_nomination
+    Ends an active nomination and returns the updated nomination.
+
+    The `unnominate_reason` field is the only allowed and required field
+    for this operation. The nomination will automatically be marked as
+    `active = false` and the datetime of this operation will be added to
+    the `unwatched_at` field.
+
+    #### Request body
+    >>> {
+    ...     'unnominate_reason': 'He needs some time to mature his Python knowledge'
+    ... }
+
+    ### Response format
+    See `GET /bot/nominations/<id:int>`
+
+    #### Status codes
+    - 200: returned on success
+    - 400: returned on failure for the following reasons:
+        - `unnominate_reason` is missing from the request body;
+        - Any other field is present in the request body;
+        - The nomination was already inactiive.
+    - 404: if an infraction with the given `id` could not be found
+    """
     serializer_class = NominationSerializer
-    queryset = Nomination.objects.prefetch_related('author', 'user')
-    frozen_fields = ('author', 'inserted_at', 'user')
+    queryset = Nomination.objects.all()
+    filter_backends = (DjangoFilterBackend, SearchFilter, OrderingFilter)
+    filter_fields = ('user__id', 'actor__id', 'active')
+    frozen_fields = ('id', 'actor', 'inserted_at', 'user', 'unwatched_at', 'active')
+    frozen_on_create = ('unwatched_at', 'unnominate_reason', 'active', 'inserted_at')
 
-    def update(self, request, *args, **kwargs):
+    def create(self, request, *args, **kwargs):
         """
-        DRF method for updating a Nomination.
+        DRF method for creating a Nomination.
 
         Called by the Django Rest Framework in response to the corresponding HTTP request.
         """
+        for field in request.data:
+            if field in self.frozen_on_create:
+                raise ValidationError({field: ['This field cannot be set at creation.']})
+
+        user_id = request.data.get("user")
+        if Nomination.objects.filter(active=True, user__id=user_id):
```

Comment:
prefer explicit `exists`, see https://docs.djangoproject.com/en/2.2/ref/models/querysets/#django.db.models.query.QuerySet.exists

Reply:
Agreed, I'll change this.